src/cmdline/cmdline_input.ts

```diff
@@ -0,0 +1,237 @@
+import { QuickPick, QuickPickItem, ThemeIcon, window, Disposable, QuickInputButton } from "vscode";
+
+import { disposeAll } from "../utils";
+import { calculateInputAfterTextChange } from "../utils/cmdline_text";
+import { createLogger } from "../logger";
+
+const logger = createLogger("CmdLineInput", false);
+
+export interface CmdlineInputListeners {
+    onAccept: () => void;
+    onChangeValue: (value: string, toInput: string) => void;
+    onChangeSelection: (selectionIndex: number) => void;
+    onHide: () => void;
+}
+
+export interface VisibilitySome {
+    visible: true;
+    level: number;
+}
+
+export interface VisibilityNone {
+    visible: false;
+}
+
+export type Visibility = VisibilitySome | VisibilityNone;
```

Comment:
Can't level just be 0 to indicate invisible? What about number | undefined?

Reply:
I could. It's a personal preference of mine to have the types be a bit explicit, but if you prefer `number?`, we can do that

Reply:
Reduce the volume of code a reader has to grok. An optional is a known pattern.